doc/components/csr.md

```diff
@@ -0,0 +1,324 @@
+
+### CsrFieldConfig
+
+The `CsrFieldConfig` contains the architectural definition of an individual field within a CSR. The `CsrFieldConfig` exposes the following functionality:
+
+- A name for the field (called `name`).
+- A starting bit position for the field within the register (called `start`).
+- A bit width for the field (called `width`).
+- An access mode for the field (called `access`).
+- An optional reset value for the field (called `resetValue`). If not provided, the reset value defaults to 0.
+- An optional list of legal values for the field (called `legalValues`).
+- A `validate()` method to check for configuration correctness and consistency.
+- A method `transformIllegalValue()` to map illegal field values to some legal one. The default implementation statically maps to the first legal value in the `legalValues` list.
+
+#### CSR Field Access Modes
+
+At the CSR field granularity, the access modes, defined in the Enum `CsrFieldAccess`, are:
+
+- Read only
+- Read/write
+- Write ones clear (read only, but writing it has a side effect)
+- Read/write legal (can only write a legal value)
+
+Note that field access rules apply to both frontdoor and backdoor accesses of the register.
+
+To support the read/write legal mode, the configuration must provide a non-empty list of legal values to check against. In the hardware's logic construction, if a write is attempting to place an illegal value in the field, this write data is remapped to a legal value per the `transformIllegalValue()` method. This method can be custom defined in a derived class of `CsrFieldConfig` but has a default implementation that can be used as is.
+
+#### Validation of CsrFieldConfig
+
+The following checks are run:
+
+- The `resetValue` must fit within the field's `width`.
+- If the field has read/write legal access, the `legalValues` must not be empty.
+- If the field has read/write legal access, the `resetValue` must appear in the `legalValues`.
+- If the field has read/write legal access, every value in `legalValues` must fit within the field's `width`.
+
+### API for Csr
+
+As the `Csr` class is an extension of `LogicStructure`, it inherits all of the functionality of `LogicStructure`. This includes the ability to directly assign and/or consume it like any ROHD `Logic`.
+
+In addition, the following attributes and methods are exposed:
+
+- Accessors to all of the member attributes of the underlying `CsrInstanceConfig`.
+- `Logic getField(String name)` which returns a `Logic` for the field within the CSR with the name `name`. This enables easy read/write access to the fields of the register if needed for logic in the parent hardware module.
+- `Logic getWriteData(Logic wd) ` which returns a `Logic` tranforming the input `Logic` in such a way as to be a legal value to write to the given register. For example, if a field is read only, the current data for that field is grafted into the input data in the appropriate bit position.
```

Comment:
I think a practical example of using `getWriteData` would be valuable here.  Is there a way we can make the backdoor interface support a write enable per-field or something so that users don't need to stack up `getWriteData` calls?